File: ios/VisePandaIOS/Models/AuthModels.swift
import Foundation

struct SupabaseAuthSession: Codable, Equatable {
    var accessToken: String
    var refreshToken: String
    var expiresIn: Int
    var tokenType: String
    var user: SupabaseAuthUser

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresIn = "expires_in"
        case tokenType = "token_type"
        case user
    }
}

struct SupabaseAuthUser: Codable, Equatable {
    var id: String
    var email: String?
}

/// Supabase's `POST /auth/v1/signup` returns two different shapes depending
/// on project settings, both HTTP 200:
/// - Email confirmation OFF: the same session shape as sign-in
///   (`access_token`/`refresh_token`/... + nested `user`).
/// - Email confirmation ON (this project's actual config, confirmed by
///   `confirmation_sent_at` in the real response): a flat user object with no
///   `access_token` at all — the traveler must click the emailed link first.
/// Decoding this response straight into `SupabaseAuthSession` throws
/// `DecodingError.keyNotFound` for `access_token`, which iOS shows verbatim as
/// "The data couldn't be read because it is missing." — a real bug, not a
/// transient network failure. This type tries the session shape first, then
/// falls back to the flat user shape so signup can tell the two apart.
enum SupabaseSignUpResult {
    case signedIn(SupabaseAuthSession)
    case confirmationRequired(email: String)
}

struct SupabaseAuthRequest: Codable {
    var email: String
    var password: String
}

struct SupabaseRefreshRequest: Codable {
    var refreshToken: String

    enum CodingKeys: String, CodingKey {
        case refreshToken = "refresh_token"
    }
}

struct SupabaseAuthError: Codable {
    var error: String?
    var errorDescription: String?
    var msg: String?
    var message: String?

    enum CodingKeys: String, CodingKey {
        case error
        case errorDescription = "error_description"
        case msg
        case message
    }

    var displayMessage: String {
        message ?? msg ?? errorDescription ?? error ?? "Authentication failed"
    }
}

File: android/app/src/main/kotlin/space/go2china/visepanda/data/model/AuthModels.kt
package space.go2china.visepanda.data.model

import com.google.gson.annotations.SerializedName

data class AuthResponse(
    @SerializedName("access_token") val accessToken: String,
    @SerializedName("token_type") val tokenType: String,
    @SerializedName("expires_in") val expiresIn: Long,
    @SerializedName("refresh_token") val refreshToken: String,
    @SerializedName("user") val user: AuthUser
)

data class AuthUser(
    @SerializedName("id") val id: String,
    @SerializedName("email") val email: String?
)

/**
 * Supabase's /auth/v1/signup returns two different shapes depending on the
 * project's "Confirm email" setting: a full session (this project has email
 * confirmation ON, so this is the disabled-confirmation fallback) or a flat
 * user object with no access_token — the latter is the success path here,
 * not a decode failure, mirrors iOS SupabaseAuthClient.SupabaseSignUpResult.
 */
sealed class SignUpOutcome {
    data class SignedIn(val response: AuthResponse) : SignUpOutcome()
    data class ConfirmationRequired(val email: String) : SignUpOutcome()
}

/** Supabase GoTrue error body shape, mirrors iOS SupabaseAuthError. */
data class SupabaseAuthError(
    @SerializedName("error") val error: String? = null,
    @SerializedName("error_code") val errorCode: String? = null,
    @SerializedName("error_description") val errorDescription: String? = null,
    @SerializedName("msg") val msg: String? = null,
    @SerializedName("message") val message: String? = null,
) {
    val displayMessage: String
        get() = if (errorCode == "over_email_send_rate_limit") {
            "Too many signup emails were sent recently. Please wait a while and try again, or sign in if you already created this account."
        } else {
            message ?: msg ?: errorDescription ?: error ?: "Authentication failed"
        }
}

data class LoginRequest(
    @SerializedName("email") val email: String,
    @SerializedName("password") val password: String
)

data class SignUpRequest(
    @SerializedName("email") val email: String,
    @SerializedName("password") val password: String
)

data class RefreshTokenRequest(
    @SerializedName("refresh_token") val refreshToken: String
)
